feat(Tile): add tile component

kmcfaul opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Closes #4524

patternfly-build commented:
PF4 preview: https://patternfly-react-pr-4610.surge.sh

tlabaj commented on packages/react-core/src/components/Tile/examples/Tile.md:
```diff
@@ -0,0 +1,110 @@
+---
```
It looks like the "extra content" example is missing.

kmcfaul commented:
@mcarrano @jessiehuff 
Re: @dlabrecq comment

I can add an `onKeyDown` handler to the tile, but which keys to we want to allow for selection? `Enter` only or both enter and `Space`?

mcoker commented on packages/react-core/src/components/Tile/Tile.tsx:
```diff
@@ -0,0 +1,53 @@
+  isSelected?: boolean;
+  /** Flag indicating if the tile is disabled */
+  isDisabled?: boolean;
+  /** Flag indicating if the tile icon is stacked */
```
Just a nit, but this might say "tile header is stacked" since that's the layout variation - though it does stack the icon, so that works, too :)

mcoker commented on packages/react-core/src/components/Tile/Tile.tsx:
```diff
@@ -0,0 +1,53 @@
+  /** Flag indicating if the tile icon is stacked */
+  isStacked?: boolean;
+  /** Flag indicating if the stacked tile icon is large */
+  isLargeIcon?: boolean;
```
I would rename this to `isDisplayLarge`. This variation could introduce other styles that don't apply to the icon. Also this is a new "display" concept we've added for more compact (application) vs larger (website/marketing) designs, which is why it would be `isDisplayLarge` instead of something like `isLarge`, which could do something different, like change the width or height or something like that and be different in concept to what `isDisplayLarge` does.

mcoker replied:
Also that matches the core class, so we will avoid name conflicts in the future since these match.

mcoker commented:
Looking at the extra content examples, the core examples limit the tile width, and place them in a flex layout, so the cards display side by side and are equal heights:

<img width="1013" alt="Screen Shot 2020-07-29 at 7 39 56 PM" src="https://user-images.githubusercontent.com/35148959/88867541-5d30b800-d1d3-11ea-84ec-cc0a19d36533.png">

Here are those examples in this PR:

<img width="844" alt="Screen Shot 2020-07-29 at 7 31 41 PM" src="https://user-images.githubusercontent.com/35148959/88867168-48075980-d1d2-11ea-965f-e351d2739111.png">

@christiemolloy I'm assuming that example in the core docs shows how cards of different heights (with differing amounts of content) should have the same height when placed in a row? If so I'm wondering if we should use a layout for that instead of custom workspace styles, and if that belongs in a demo instead? WDYT?

christiemolloy commented:
@mcoker thats a good point, without the workstyle style of `display: flex` the cards wont all grow to the same height. I think a demo with a flex layout could definitely solve this though

mcarrano commented:
@kmcfaul I'm going to pull together a quick meeting to discuss what the intended behavior should be. Tiles can be used like a checkbox or radio button group. Not sure that the examples currently illustrate that.

kmcfaul commented:
@mcoker @christiemolloy 
Updated `isStacked` property description.
Renamed `isLargeIcon` to `isDisplayLarge`.
Updated `Tile with extra content` to use `Flex`.

@mcarrano 
I'm not sure the Tile component examples would show this behavior, it sounds more like a demo.

kmcfaul commented:
@mcarrano @jessiehuff @maryshak1996

Added two tile demos to the demos section: single select allowing `space` as a selection key, and multiple select allowing `enter` as a selection key.

There is a bit of weirdness with the mouse, because the event can pick up the title div as a target instead of the outer div where the `id` is present. I'm still thinking this one through, but Tile may have to pipe down the onClick to the title div (but that would lock in using `id` or another data property for comparison).

Another point of confusion is that the styles for the focus and selected states are very similar, so it can be hard to tell what is selected versus what is focused. @mcoker

kmcfaul commented:
@redallen Any idea what the build error is pointing to? Docs builds fine locally.

kmcfaul commented:
@jessiehuff I added a note and link to the demo section in the Tile.md, let me know if that looks okay. Eventually, the demo section will live in a tab alongside the rest of the component docs, at which point it may no longer be necessary.
